db.go

```diff
@@ -469,6 +476,12 @@ func (db *DB) mmap(minsz int) (err error) {
 		return err
 	}
 
+	if db.MaxSize != 0 && fileSize < size && db.MaxSize < size {
```

Comment:
```suggestion
	if db.MaxSize != 0 && db.MaxSize < size {
```

Reply:
When opening the file in the beginning, the `minsz` is `options.InitialMmapSize`, which is set by users and may be a huge value which is far larger than the actual file size. In that case, it will wrongly return `berrors.ErrMaxSizeReached` if MaxSize isn't 0.

Reply:
> If a user does that, on Windows, the behaviour is that the file is expanded to `size`

I don't think it's true based on my understanding, and also [createfilemappinga doc](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createfilemappinga). No matter what's size of the file mapping object, the file real size won't expand until bbolt really writes data beyond the file end offset. It should be easy to write a simple test to verify. But the following calling of `WriteAt` might increase the file real size; which makes it hard to control the real size.

For non-windows platform, mmap won't expand the file size either.

Reply:
this function calls mmap here: https://github.com/etcd-io/bbolt/blob/main/db.go#L492
the windows implementation calls Truncate: https://github.com/etcd-io/bbolt/blob/main/bolt_windows.go#L71 which changes the size of the file

Reply:
OK, `Truncate` isn't allowed on a file with a user-mapped section open on Windows platform, it's exactly the reason why bbolt needs to `Truncate` before mapping the file. FYI. https://github.com/etcd-io/bbolt/issues/378#issuecomment-1378121966

For non-Windows platform, my above comment https://github.com/etcd-io/bbolt/pull/929/files#r2031890550 still stands.

Reply:
I'm going to add a test to hit this behaviour where the user sets a high InitialMmapSize on a database that is already above the limit, to make sure there aren't unintentional consequences on linux

Reply:
I think this means I will also need to move this oversize check into the windows-specific logic

Reply:
I've done the changes I contemplated above. I moved the mmap check into the windows-specific code. I also added logic to make sure that mmaps that aren't going to change the size of the file are still allowed, even if it's larger than the max size